### PR TITLE
Add type tests and fix return type edge cases

### PR DESCRIPTION
**Pull request by epeli**
Thanks for creating this library! This is so far the best way to handle possibly undefineds in TypeScript I've found.

And sorry for such an invasive pull request. I found few edge cases with return types including #19. This PR adds type test for those using [conditional-type-checks](https://github.com/dsherret/conditional-type-checks) and fixes those issues.

Unfortunately I had to remove the `include` config from tsconfig.json in order to expose those issues in tests. This meant that I had to rework the build script a bit. As a side effect cleaned up the published package. Ie. test files are not shipped to npm anymore as they [are now](https://unpkg.com/ts-optchain@0.1.3/dist/__tests__/).

### CODE REVIEW

File: __tests__/types.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+assert<Has<typeof resUnion, undefined>>(false);
+assert<Has<typeof resUnion, null>>(false);
+
+
+const resNoDefault = oc(x).a.b();
+// Has string and undefined
+assert<Has<typeof resNoDefault, string>>(true);
+assert<Has<typeof resNoDefault, undefined>>(true);
+
+const resExists = oc(x).exists();
```
Comment (wielrls) [inline __tests__/types.ts]: We have historically taken the more conservative approach of assuming that even if a property is marked required on an interface definition, we don't take this fact for granted at runtime. The context is: we have typically used `oc` for scenarios where we do not have 100% guarantees at runtime about the structure of processed data, despite our most optimistic hopes at development time (e.g., dealing with a misbehaving API, an inconsistently implemented API, etc).

Comment (epeli) [reply __tests__/types.ts]: Yes, I agree it makes sense. I'll revert this.

Comment (epeli) [reply __tests__/types.ts]: Rebased the PR for this.

**Comment by epeli** [comment]
Added few more tests to this :)

**Comment by epeli** [comment]
Hi @wielrls is there something you would like me to do before this can be merged? Thanks.

File: __tests__/types.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
+/**
```
Comment (wielrls) [inline __tests__/types.ts]: Good call adding these unit tests specifically focused on type definitions.

File: package.json [hunk 3]
```diff
@@ -8,17 +8,22 @@
   },
   "main": "./dist/index.js",
   "types": "./dist/index.d.ts",
+  "files": [
+    "dist"
```
Comment (wielrls) [inline package.json]: 👍

File: src/index.ts [hunk 4]
```diff
@@ -36,10 +31,10 @@ export interface IDataAccessor<T> {
   (defaultValue: NonNullable<T>): NonNullable<T>;
 
   /**
-   * Data accessor with null default value.
-   * @param defaultValue
+   * Data accessor without a default value. If no data exists,
+   * `undefined` is returned.
    */
-  (nullDefaultValue: T extends null ? null : never): Defined<T>;
+  (): T | undefined
```
Comment (wielrls) [inline src/index.ts]: This change breaks the test case mentioned above.

File: __tests__/types.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+declare const x: X;
+
+const resWithDefault = oc(x).a.b("");
+assert<IsExact<typeof resWithDefault, string>>(true);
+
+const resMaybeNull = oc(x).a.maybeNull("")
+assert<IsExact<typeof resMaybeNull, string>>(true);
+
```
Comment (wielrls) [inline __tests__/types.ts]: Can you add a test for the previously supported case of setting default value to `null`:
```typescript
const resMaybeNullWithDefaultNull = oc(x).a.maybeNull(null);
assert<Has<typeof resMaybeNullWithDefaultNull, string>>(true);
assert<Has<typeof resMaybeNullWithDefaultNull, null>>(true);
```

File: src/index.ts (hunk 4 above)
Comment (epeli) [reply src/index.ts]: Heh, this is why type tests are important too 😉

File: __tests__/types.ts (hunk 5 above)
Comment (epeli) [reply __tests__/types.ts]: Unfortunately it seems to be very hard to support all these cases.

If I revert the implementation to your original the null test passes, but the union test fails:

image

Which is in my opinion a lot more important because I can see uses cases where I'd want to remove undefined from a union. It's actually why I created this PR (#19 ). Removing the string from `string | null` union seems really weird to me.

The however the implementation can be modified like this:

image

But it won't still pass completely

image

I think the implementation in the PR is the least bad one.

Comment (wielrls) [reply __tests__/types.ts]: Ah right! I ran into this when working on https://github.com/rimeto/ts-transform-optchain. The workaround I found:
```
export interface IDataAccessor<T> {

  /**
   * Data accessor without a default value. If no data exists,
   * `undefined` is returned.
   */
  (noDefaultValue?: undefined): Defined<T> | undefined;

...
```

See if this fixes the problem?

Comment (epeli) [reply __tests__/types.ts]: Doesn't seem to :/

Comment (wielrls) [reply __tests__/types.ts]: weird -- it appears to work on my checkout of your branch using ts 3.3.3333:
```diff
--- a/src/index.ts
+++ b/src/index.ts
@@ -23,6 +23,11 @@ export type Defined<T> = Exclude<T, undefined>;
 ////////////////////////////
 
 export interface IDataAccessor<T> {
+  /**
+   * Data accessor without a default value. If no data exists,
+   * `undefined` is returned.
+   */
+  (noDefaultValue?: undefined): Defined<T> | undefined;
 
   /**
    * Data accessor with default value.
@@ -34,7 +39,7 @@ export interface IDataAccessor<T> {
    * Data accessor without a default value. If no data exists,
    * `undefined` is returned.
    */
-  (): T | undefined
+  (nullDefaultValue: T extends null ? null : never): Defined<T>;
 }
 
 ///////////////////////////
```

Comment (epeli) [reply __tests__/types.ts]: Ah missunderstood your patch previously. All green now! 🎉
